### PR TITLE
WIP: Fully support shift_len > 0 in clustering.distance_matrix

### PR DESCRIPTION
**Pull request by flixha**
### What does this PR do?
Issue #451 describes how `utils.clustering.distance_matrix` did not properly handle `shift_len > 0`, and this PR implements an attempt to solve that.

### Why was it initiated?  Any relevant Issues?
Here's a first attempt that may need a bit discussion, as there is one important point that turned up:
- Previously, by providing `shift_len`, not only were template streams shifted against one another as a whole, but each trace of each template was able to be shifted. The maximum CC sum then reflected the best correlation that can be achieved by doing that. That wasn't fully obvious to me in the discussion in #451. Of course that has a big effect on how a matrix of the positions can look like that we want to have returned. Then there's basically a 2-D position / shift matrix for each template, and the size of that matrix is different for each template.

For now, I have implemented the following solution:

- the parameter `allow_individual_trace_shifts` controls whether traces can be shifted individually in `distance_matrix` and `cross_chan_correlation`
- If `allow_individual_trace_shifts=False`, then the shift-matrix has the same size as the distance-matrix (`n_templates x n_templates`)
- If `allow_individual_trace_shifts=True`, then what is returned as shift-matrix is a 3-D matrix with size `n_templates x n_templates x n_max_traces`, where e.g., `shift_mat[0,1,:]` gives the shifts for each trace when comparing template 0 with template 1.
- the order in which trace-shifts appear in shift_mat is defined by the order in which the traces appear in each template
- for a template with `n_traces < n_max_traces`, the shift-values for indices greater `n_traces` are NaNs.

All in all, the above point meant that there's lots of slightly ugly array-indexing in the code now that may not always be super easy to read. Let me know whether this solution is along the lines how you could imagine it being most useful in the future, or whether you think we should change what the functions return now - I'm sure there's room for improvement here!

### PR Checklist
- [X] `develop` base branch selected?
- [X] This PR is not directly related to an existing issue (which has no PR yet).
- [X] All tests still pass.
- [X] Any new features or fixed regressions are covered via new tests.
- [ ] Any new or changed features are fully documented.
- [ ] Significant changes have been added to `CHANGES.md`.
- [ ] ~~First time contributors have added your name to `CONTRIBUTORS.md`.~~

For demonstration, the test from #451 with the new options:
```python
import glob, os
from obspy import read
from eqcorrscan import tests
from eqcorrscan.utils.clustering import distance_matrix

testing_path = os.path.join(os.path.dirname(tests.__file__),
                            'test_data', 'WAV', 'TEST_')
stream_files = glob.glob(os.path.join(testing_path, '*DFDPC*'))[0:10]
stream_list = [read(stream_file) for stream_file in stream_files[0:4]]
#stream_list = [st.select(station='EORO', channel='SHZ') for st in stream_list]  #single-trace test
stream_list = [st.select(sampling_rate=200) for st in stream_list]
dist_mat, shift_mat, shift_dict = distance_matrix(
    stream_list=stream_list, cores=1, shift_len=0.5, allow_individual_trace_shifts=False)
dist_mat2, shift_mat2, shift_dict = distance_matrix(
    stream_list=stream_list, cores=1, shift_len=0.5, allow_individual_trace_shifts=True)
```

### CODE REVIEW

**Comment by flixha** [comment]
I may also want to return the shift-matrix when running  `clustering.cluster`; testing that tomorrow.

File: eqcorrscan/utils/clustering.py [hunk 1]
```diff
@@ -61,6 +69,8 @@ def cross_chan_correlation(st1, streams, shift_len=0.0, xcorr_func='fftw',
     """
     # Cut all channels in stream-list to be the correct length (shorter than
     # st1 if stack = False by shift_len).
+    allow_individual_trace_shifts =\
+        allow_individual_trace_shifts and shift_len > 0
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/clustering.py:73]:
```suggestion
    allow_individual_trace_shifts = (
        allow_individual_trace_shifts and shift_len > 0)
```

File: eqcorrscan/utils/clustering.py [hunk 2]
```diff
@@ -79,32 +89,46 @@ def cross_chan_correlation(st1, streams, shift_len=0.0, xcorr_func='fftw',
         # We need to copy it first.
         streams = [stream.copy() for stream in streams]
     # Check which channels are in st1 and match those in the stream_list
-    st1, streams, stream_indexes = _prep_data_for_correlation(
+    st1, prep_streams, stream_indexes = _prep_data_for_correlation(
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/clustering.py:92]: I like this - renaming to `prep_streams` - if you want you could also rename `st1` to something else to make it clear that `st1` and the output of this are different.

Comment (calum-chamberlain) [inline eqcorrscan/utils/clustering.py:92]:
```suggestion
    st_preped, prep_streams, stream_indexes = _prep_data_for_correlation(
```

File: eqcorrscan/utils/clustering.py [hunk 3]
```diff
@@ -79,32 +89,46 @@ def cross_chan_correlation(st1, streams, shift_len=0.0, xcorr_func='fftw',
...
         stream=st1.copy(), templates=streams,
         template_names=list(range(len(streams))), force_stream_epoch=False)
     # Run the correlations
     multichannel_normxcorr = get_stream_xcorr(xcorr_func, concurrency)
     [cccsums, no_chans, _] = multichannel_normxcorr(
-        templates=streams, stream=st1, cores=cores, stack=False, **kwargs)
+        templates=prep_streams, stream=st1, cores=cores, stack=False, **kwargs)
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/clustering.py:98]:
```suggestion
        templates=prep_streams, stream=st_preped, cores=cores, stack=False, 
        **kwargs)
```

File: eqcorrscan/utils/clustering.py [hunk 4]
```diff
@@ -119,6 +143,11 @@ def distance_matrix(stream_list, shift_len=0.0, cores=1):
         matrix for
     :type shift_len: float
     :param shift_len: How many seconds for templates to shift
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/clustering.py:145]:
```suggestion
    :param shift_len:
        Seconds to shift the streams by (total value for negative and positive
        direction together)
```
To match above, which is clearer than my old version!

File: eqcorrscan/utils/clustering.py [hunk 5]
```diff
@@ -133,19 +162,46 @@ def distance_matrix(stream_list, shift_len=0.0, cores=1):
     .. note::
         Requires all traces to have the same sampling rate and same length.
     """
+    allow_individual_trace_shifts =\
+        allow_individual_trace_shifts and shift_len > 0
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/clustering.py:166]:
```suggestion
    allow_individual_trace_shifts = (
        allow_individual_trace_shifts and shift_len > 0)
```

File: eqcorrscan/utils/clustering.py [hunk 6]
```diff
@@ -133,19 +162,46 @@ def distance_matrix(stream_list, shift_len=0.0, cores=1):
...
+        # get the shortest distance for each correlation pair
+        dist_mat_shortest = np.minimum(dist_mat, dist_mat.T)
+        # Indicator says which matrix has shortest dist: value 0: mat2; 1: mat1
+        mat_indicator = dist_mat_shortest == dist_mat
+        mat_indicator = np.repeat(mat_indicator[:, :, np.newaxis],
+                                  n_shifts_per_stream, axis=2)[:, :]
+        # Get shift for the shortest distances
+        shift_mat = shift_mat[:, :, 0:n_shifts_per_stream][:, :]
+        shift_mat = shift_mat * mat_indicator +\
+            np.transpose(shift_mat, [1, 0, 2]) * (1 - mat_indicator)
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/clustering.py:201]:
```suggestion
        shift_mat = (
            shift_mat * mat_indicator +  
            np.transpose(shift_mat, [1, 0, 2]) * (1 - mat_indicator))
```

File: eqcorrscan/tests/clustering_test.py [hunk 7]
```diff
@@ -84,10 +84,32 @@ def test_distance_matrix(self):
         for st in stream_list:
             for tr in st:
                 tr.data = tr.data[0:shortest_tr]
-        dist_mat = distance_matrix(stream_list=stream_list, cores=4)
+        # Test with zero shift-length
+        dist_mat, shift_mat = distance_matrix(stream_list=stream_list, cores=4)
         self.assertEqual(dist_mat.shape[0], len(stream_list))
         self.assertEqual(dist_mat.shape[1], len(stream_list))
 
+        # Test with some shift-length, but only shift templates as a whole
```
Comment (calum-chamberlain) [inline eqcorrscan/tests/clustering_test.py:92]: Can you make this a separate test please? It will help with debugging, and will allow it to run concurrently to other tests.

File: eqcorrscan/tests/clustering_test.py [hunk 8]
```diff
@@ -84,10 +84,32 @@ def test_distance_matrix(self):
...
+            stream_list=stream_list, cores=4, shift_len=0.2,
+            allow_individual_trace_shifts=False)
+        self.assertEqual(dist_mat.shape[0], len(stream_list))
+        self.assertEqual(dist_mat.shape[1], len(stream_list))
+        self.assertEqual(np.array(dist_mat == dist_mat.T).all(), True)
+        self.assertEqual(shift_mat.shape[0], len(stream_list))
+        self.assertEqual(shift_mat.shape[1], len(stream_list))
+        self.assertEqual(np.array(shift_mat == shift_mat.T).all(), True)
+
+        # Test with shift_len that individual traces are allowed to shift by
```
Comment (calum-chamberlain) [inline eqcorrscan/tests/clustering_test.py:103]: As above, can this be a separate test please.

**Review by calum-chamberlain — CHANGES_REQUESTED** [review]
This looks great to me - I had some minor suggestions for changes, including nit-picking over not using `\`, I don't know why I don't like it, but for some reason I don't, sorry!

**Comment by calum-chamberlain** [comment]
I'm fairly keen for this to go in pretty much as-is for 0.4.3 - this is a great bug-fix, and the rest of your suggestions sound more like (very useful) features. I'm keen to get 0.4.3 out this week, so maybe the other bits can wait?

**Comment by flixha** [comment]
> This looks great to me - I had some minor suggestions for changes, including nit-picking over not using `\`, I don't know why I don't like it, but for some reason I don't, sorry!

Thanks for reviewing! I will fix your suggestions, and I'm totally fine with going with parentheses instead of the backslash ;-).

Regarding what the functions should return for the shifts: I was reflecting  a bit more and I find that the 3-D shift matrix that I return right now for `allow_individual_trace_shifts=True` is quite confusing, and it may be more appropriate to have 3 values to be returned then: distance-matrix, shift-matrix (with average shifts in case individual traces can shift), and a dictionary with the master template name, correlated template names, and the template-specific trace-shift matrices. The latter matrix can then vary in size for each master template.

**Comment by calum-chamberlain** [comment]
I do like the sounds of the dictionary if it is not too much extra work to code up. An extra note/example in the docstring might be useful as well.

**Comment by flixha** [comment]
> I do like the sounds of the dictionary if it is not too much extra work to code up. An extra note/example in the docstring might be useful as well.

I haven't included the dictionary yet, because some other things needed to be fixed first. I was running into some special cases where the indexing wasn't happening as intended. Compared to a dictionary, the 3-D matrix is quite useful to check the symmetry, so it's possible that in the end returning both may be the way to go.

This needs a bit more work, but I'll try to get it done for 0.4.3 this week.

**Comment by calum-chamberlain** [comment]
Hey @flixha how is this going? Is it ready for another review? Your tests look like they are passing, but the NCEDC timeouts continue to result in failed tests.

**Comment by flixha** [comment]
It's not yet ready, but I'm on it today. I noticed the failure of a client during the tests, but I was lucky at one point in the morning where the tests were able to download the data after all. But then it stopped working again.
I'll post an update here once this is ready...

**Comment by flixha** [comment]
> Hey @flixha how is this going? Is it ready for another review? Your tests look like they are passing, but the NCEDC timeouts continue to result in failed tests.

Now this is ready for another review :-). `distance_matrix` now has three returns; distance matrix, symmetric shift matrix, and the shift-dictionary. I fixed several things related to some edge cases. Let me know if you find anything to be improved. I amended the example code in the top message so that it can be used for testing this straight away.

**Review by calum-chamberlain — APPROVED** [review]

**Comment by calum-chamberlain** [comment]
This looks good to me - the test fails are still NCEDC related, but the relevant tests pass on CI and locally for me.

Are you happy for me to merge this?

**Comment by flixha** [comment]
Thanks for reviewing again. Yes, I'm happy with this being merged now!
